### PR TITLE
Cohort Diagnostics in Hydra

### PR DESCRIPTION
**Pull request by gowthamrao**

### CODE REVIEW

**Comment by gowthamrao** [comment]
@schuemie  could you please guide me on how to create the array structure required for Hydra

https://github.com/OHDSI/Hydra/pull/17/commits/03ad026387f255bb18a11933dfbf5bc8eee5564f#diff-e2a66c57bb653a2c32e657da50b8c56caae4c6433a8ddeb1ae0ba85282f62a20R31

I am getting this error
```

Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl,  :
  java.lang.ClassCastException: org.json.JSONObject cannot be cast to org.json.JSONArray
```

**Comment by schuemie** [comment]
I recommend taking a look at the actual JSON you're producing, and comparing that to the JSON that normally comes our of ATLAS. (There's example JSON [here](https://github.com/OHDSI/Hydra/blob/master/extras/ExamplePleSpecs.json)). If you did, you'd notice that the JSON you're producing in R writes your dataframe as an array of columns, whereas the ATLAS-generated JSON has arrays of rows. 

RJSONIO is great for some things, but generating JSON is not one of them. If you use jsonlite instead, that will generate arrays of rows (no need to call `as.list()`).

**Comment by gowthamrao** [comment]
Thank you for the tip @schuemie  . I am now getting the row wise array, but i still get the error below

```

Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl,  : 
  java.lang.ClassCastException: java.base/java.lang.String cannot be cast to org.json.JSONArray
```

I compared the json structure that you showed as example, to the output of jsonlite i am getting now - they look similar to me.

**Comment by schuemie** [comment]
I'm sure you'll figure it out!

**Comment by gowthamrao** [comment]
Ok @schuemie  i will try :)

Now i have this error to solve
```

Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl,  : 
  org.json.JSONException: JSONObject["separator"] not found.
```

File: extras/TestCohortDiagnostics.R [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        cohortDefinition <-
+                ROhdsiWebApi::getCohortDefinition(cohortId = cohortId, baseUrl = baseUrl)
+        df <- tidyr::tibble(
+                id = cohortId,
+                createdDate = webApiCohorts$createdDate[[i]],
+                modifiedDate = webApiCohorts$modifiedDate[[i]],
+                description = webApiCohorts$description[[i]],
+                name = stringr::str_trim(stringr::str_squish(cohortDefinition$name)),
+                expression = cohortDefinition$expression %>% 
+                        RJSONIO::toJSON(digits = 23),
```
Comment (anthonysena) [inline extras/TestCohortDiagnostics.R:27]: I think you can skip the "toJSON" conversion for the expression here as it has the JSON you need, correct?

File: extras/TestCohortDiagnostics.R [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                sql = ROhdsiWebApi::getCohortSql(cohortDefinition = cohortDefinition$expression,
```
Comment (anthonysena) [inline extras/TestCohortDiagnostics.R:28]: Perhaps consider using CirceR for this step?

`
CirceR::buildCohortQuery(CirceR::cohortExpressionFromJson(cohortDefinition$expression), options = CirceR::createGenerateOptions())
`

File: extras/TestCohortDiagnostics.R [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+cohortDefinitionsArray <- list()
+for (i in (1:nrow(cohortDefinitions))) {
+        cohortDefinition <- cohortDefinitions[i,]
+        cohortDefinitionsArray[[i]] <- list(id = cohortDefinition$id,
+                                            name = cohortDefinition$name,
+                                            createdDate = cohortDefinition$createdDate,
+                                            modifiedDate = cohortDefinition$modifiedDate,
+                                            expression = cohortDefinition$expression %>% 
+                                                    RJSONIO::fromJSON(digits = 23))
```
Comment (anthonysena) [inline extras/TestCohortDiagnostics.R:48]: Generally speaking, I'd avoid using any of the R JSON packages when working with the circe expressions as they tend to change the structure which then causes issues.

File: extras/TestCohortDiagnostics.R [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+specifications <- specifications %>% 
+        jsonlite::toJSON(pretty = TRUE) 
+
+
+packageFolder <- "c:/temp/hydraOutput/CohortDiagnostics"
+unlink(packageFolder, recursive = TRUE)
+Hydra::hydrate(specifications = specifications, 
+               outputFolder = packageFolder, 
+               skeletonFileName = "D:\\git\\github\\gowthamrao\\Hydra\\inst\\skeletons\\CohortDiagnosticsStudy_v0.0.1.zip", 
+               packageName = "eunomiaTemp")
```
Comment (anthonysena) [inline extras/TestCohortDiagnostics.R:66]: Should this source the skeleton from the Hydra package since you've included this in the branch?

**Comment by gowthamrao** [comment]
I changed the code to simply it (so that a simple cohort may be used with hydra)

> Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl,  : 
  java.lang.RuntimeException: org.json.JSONException: JSONObject["skeletonType"] not a string.

**Comment by gowthamrao** [comment]
For some reason 'skeleton type' is not being recognized

> Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl, :
> java.lang.RuntimeException: org.json.JSONException: JSONObject["skeletonType"] not a string.
> 

If i point to the skeleton package in function call as follows

```
 Hydra::hydrate(specifications = specifications, 
                outputFolder = packageFolder, 
                skeletonFileName = "D:\\git\\github\\gowthamrao\\Hydra\\inst\\skeletons\\CohortDiagnosticsStudy_v0.0.1.zip", 
                packageName = "eunomiaTemp")
```

i  get

> Warning: Output folder c:/temp/hydraOutput/CohortDiagnostics already exists. Any contents will be overwritten
> Error in .jcall("RJavaTools", "Ljava/lang/Object;", "invokeMethod", cl,  : 
>   org.json.JSONException: JSONObject["separator"] not found.

**Comment by gowthamrao** [comment]
Updated code and tested to work! @schuemie could you please review

**Comment by schuemie** [comment]
I can now hydrate the skeleton, but I have some comments:

About the example code:
- Please move the code for generating the JSON to whatever is generating the JSON. Only include a full example JSON (like the PLE example) that already has the cohort definitions embedded.

About the skeleton:
- The skeleton doesn't use renv, but this is recommended practice for skeletons
- The README file doesn't conform to the README files of the other skeletons (for one, please use the name 'README.md' with those exact capitals)
- The CodeToRun.R after hydration refers to 'Skeleton::', which is not an existing package.
- Why does CodeToRun.R require `library(CohortDiagnostics)`? This should just be dependency of the new study package. For example, the PLE skeleton does not require `library(CohortMethod)`.
- Defaulting the output folder to a subfolder of the git repo (as done in CodeToRun.R) seems like not something we'd recommend for users, and is inconsistent with the other package skeletons.
- Why does the `runCohortDiagnostics()` function require the user to specify the package name? This seems a recipe for mistakes. The package name should be hard-coded and set by Hydra.
- Please add a `uploadResults()` function (like in the PLE skeleton) so people can upload the results when executing as part of a network study.
- Deleting the output folder if it exists (as currently in CodeToRun.R) seems like maybe not the best idea?
- I would think that the option to `runInclusionStatistics`, `runCohortOverlap` , etc. would be specified in the JSON, and not something that the study-package user should decide?

Remember, the main use case for a Hydra skeleton is: 
- Study lead designs a study.
- Study lead hydrates the skeleton to a full study package based on the design and puts it on ohdsi-studies.
- Study lead requests data partners to run the study package.
- A tech person (who perhaps knows nothing about the study) at the data partner site downloads the study package, fills in the arguments required to connect to the database, and runs the package. Then uploads the results to the SFTP server.
- The study leads fetches the results from the SFTP server and synthesizes them across sites.

I haven't yet tested the hydrated skeleton. Will do after these changes have been made.

**Comment by gowthamrao** [comment]
@schuemie  i might need some help with renv. I am getting this message

`Error: failed to retrieve package 'DatabaseConnector' `

with trace showing 
```
9.
stop(sprintf(fmt, ...), call. = call.) 
8.
stopf("failed to retrieve package '%s'", record$Package) 
7.
renv_retrieve_repos(record) 
6.
renv_retrieve_impl(package) 
5.
handler(package, renv_retrieve_impl(package)) 
4.
renv_retrieve(packages) 
3.
renv_restore_run_actions(project, diff, current, lockfile, rebuild) 
2.
restore(project = project, library = libpaths, prompt = FALSE) 
1.
renv::init() 

```

**Comment by schuemie** [comment]
@gowthamrao : Could you move the skeleton to the new repo? https://github.com/OHDSI/SkeletonCohortDiagnostics

**Comment by gowthamrao** [comment]
@schuemie  done - it is here https://github.com/OHDSI/SkeletonCohortDiagnostics

What do we do next? Are you planning to update the Hydra package to pick up directly from the Skeleton repo? Do we use the same naming convention for the skeleton zips to identify the skeleton versions? Will there be a parameter in hydra to find the right skeleton?

I would also recommend that we retain the ability to pick up local skeletons and hydrate them

**Comment by schuemie** [comment]
I meant the unzipped skeleton. See [SkeletonComparativeEffectiveness](https://github.com/OHDSI/SkeletonComparativeEffectStudy) for an example

**Comment by gowthamrao** [comment]
That makes more sense. Done.

**Comment by schuemie** [comment]
Thanks! This makes it easier to discuss. Regarding your question on DatabaseConnector, note that you're referencing a version that is not yet on CRAN, but are telling renv to get it from CRAN: https://github.com/OHDSI/SkeletonCohortDiagnostics/blob/main/renv.lock#L444

This version will be released to CRAN in the coming days

**Pull request merged**